src/policy/markcompactspace.rs

```diff
@@ -318,69 +321,126 @@ impl<VM: VMBinding> MarkCompactSpace<VM> {
         Self::is_marked(object)
     }
 
-    pub fn calculate_forwarding_pointer(&self) {
-        let start = self.common.start;
-        let end = self.pr.cursor();
-        let mut to = start;
-
-        let linear_scan =
-            crate::util::linear_scan::ObjectIterator::<VM, MarkCompactObjectSize<VM>, true>::new(
-                start, end,
-            );
-        for obj in linear_scan.filter(|obj| Self::to_be_compacted(*obj)) {
-            let copied_size =
-                VM::VMObjectModel::get_size_when_copied(obj) + Self::HEADER_RESERVED_IN_BYTES;
-            let align = VM::VMObjectModel::get_align_when_copied(obj);
-            let offset = VM::VMObjectModel::get_align_offset_when_copied(obj);
-            to = align_allocation_no_fill::<VM>(to, align, offset);
-            let new_obj = VM::VMObjectModel::get_reference_when_copied_to(
-                obj,
-                to + Self::HEADER_RESERVED_IN_BYTES,
-            );
-
-            Self::store_header_forwarding_pointer(obj, new_obj);
-
-            trace!(
-                "Calculate forward: {} (size when copied = {}) ~> {} (size = {})",
-                obj,
-                VM::VMObjectModel::get_size_when_copied(obj),
-                to,
-                copied_size
-            );
+    /// Linear scan all the live objects in the given memory region
+    fn linear_scan_objects(
+        &self,
+        range: Range<Address>,
+        to_be_compacted_only: bool,
+    ) -> impl Iterator<Item = ObjectReference> {
+        let linear_scan = crate::util::linear_scan::ObjectIterator::<
+            VM,
+            MarkCompactObjectSize<VM>,
+            true,
+        >::new(range.start, range.end);
+        linear_scan.filter(move |obj| !to_be_compacted_only || Self::to_be_compacted(*obj))
+    }
+
+    /// Iterate over all contiguous memory regions in this space.
+    /// For contiguous space, this iterator should yield only once, and returning a contiguous memory region covering the whole space.
+    fn iterate_contiguous_regions(&self) -> impl Iterator<Item = (Address, Address)> + '_ {
```

Comment:
I think this method can be moved to `MonotonePageResource`. Also there is a method that looks very similar in `MonotonePageResource`. Does `for_allocated_regions` work for this case?
https://github.com/mmtk/mmtk-core/blob/2e29b978272cee30be85cb1fedb7a303086aa02e/src/util/heap/monotonepageresource.rs#L298

If they are doing the same thing, I suggest we keep one implementation (either the old one or the new one).

Reply:
These are basically the same thing. However, within the `MarkCompactSpace::compact`, it has to be an iterator. The function needs to bump the to-space region iterator conditionally instead of blindly iterating over the whole heap.

I removed `for_allocated_regions` and rewrote that to a function returning an iterator, so it can work on both MarkCompactSpace and ImmortalSpace (which is the only caller of this function).